### ISSUE TITLE
display backend

### ISSUE THREAD

**Issue by JarbasAI** [MycroftAI/mycroft-core#650]
similiar to forslunds AudioBackend idea https://github.com/MycroftAI/mycroft-core/pull/433

it would be good to have a mechanism that would listen for messages to display pictures, each user should be able to choose where and how to display a picture

it would be useful to either use different libs (opencv / gtk) or show in different places (living room tv screen, laptop screen, tablet...) 

an initial idea i've been playing with: https://github.com/JarbasAI/mycroft---display-backend

**Comment by penrods** [comment]
I agree, this is very similar to the idea of the audio backend.  For example, when I say "show me the weather forecast", whatever voice access point I use should have an associated display that it uses by default.  The display may be on the unit itself (i.e. a refrigerator with a high res screen built in to it) or just a visible screen nearby (i.e. a coffee pot with only a mic/speaker, but which has been configured by the user so it knows there is a nearby Roku that is connected to the kitchen TV and can be used to render visualizations).

I will be expanding the skill framework to allow a handler to respond with several types of payloads (not mutually exclusive).  This will include:
   * text (likely rendered by TTS to the user)
   * display_text (maybe different than text, e.g. "1/4 cup" instead of the spoken "one quarter of a cup")
   * audio (e.g. a URL pointing to an MP3)
   * image (standard bitmap formats)
   * html (either html code in a string or a URL)

Video will likely be supported someday, but I think HTML is adequate for now.

The enclosure would extract the appropriate piece(s) to render.  For example a Skill for "what's the forecast" might return text "59 degrees", display text "59° F", and a URL back to the skill for rendering a 5 day forecast.  On the Mark 1 it will likely only draw the "59° F" on the matrix faceplate and speak "59 degrees".  A car could speak "59 degrees" and launch a browser that shows the 5 day forecast on the in-dash screen.

And like the audio service, screens are a limited resource.  So we need a manger to control it.

Side note:  Another important piece in the long run is that the Skill will not directly control the display_service.  The skill system will pick with response to implement, potentially from several skills.  Imagine you have three weather services, one that knows it works works well in the US but only so-so worldwide, one that knows it works works well in Europe, and another that works only in China.  The phrase 'get me the forecast in XXX' might be handled by two or three of these, but depending on the location requested the result confidence would determine which one should actually be presented to the user.


TL;DR - Yes, we need to implement this.

**Comment by JarbasAI** [comment]
A service/process to listen for display requests should be implemented and parse from utterance if any backend was selected by user, if not locate source of info and display there / fallback to default backend

if we make this monitor a services folder (like skills currently) it can load any backend and its just a matter of the manager to direct messages to correct backend, allowing for any kind of backends to be implemented, a backend is not just about location, we should to be able to select which lib / strategy it will use to display

so display backends should have:
- type <-  backend service to use (library / data type to use )
- uri <- for remote acess
- name <- so user can request "kitchen" if he wants and we just look up the url and type

this is all something audioservice already does

**Comment by krisgesling** [comment]
Hey, presuming this can be closed with the inclusion of the GUI backend. Let me know if I'm misreading.

**Comment by JarbasAl** [comment]
the implementation itself certainly wont be used, for historical reasons here is what it looked like

https://www.youtube.com/watch?v=y9rb1wKZiUs

might be worth talking about multiple guis, currently we assume 1 gui (or any number of guis) displaying the same thing, this early approach allowed

"youtube metallica in the kitchen"
"show a picture in living room"
"play sports channel in the TV"

and this is not possible in the current paradigm.  But this can be discussed into a separate issue, since now its an enhancement of the gui and not a new feature
